bench, doc: benchmarking updates and fixups

jonatack opened the pull request:
Fixups and updates I noticed while writing benchmarks for #22284.

jonatack committed:
doc: update doc/benchmarking.md

jonatack committed:
bench: bench.h fixes and improvements

jonatack commented on src/bench/bench.h:
```diff
@@ -18,16 +18,18 @@
     ... do any setup needed...
-    nanobench::Config().run([&] {
+
+    bench.run([&] {
```
Could mention options here if reviewers think it would be helpful.

```cpp
bench.run([&] {

...or with options...

bench.batch(NUM).epochs(NUM).epochIterations(NUM).unit(STRING).run([&] {
```

jarolrod replied:
I do think this would be helpful to document here

jonatack replied:
Thanks for the feedback!  Done, and fixed up the help printed by `src/bench/bench_bitcoin -?`

kiminuo commented:
Concept ACK

Thanks for improving this.

jarolrod reviewed (commented):
Concept ACK, is obviously an improvement to the docs

jonatack committed:
bench: bench_bitcoin.cpp help fixups

- remove unneeded strprintf
- consistent punctuation (no EOL periods)
- sort helps by order they are printed (alphabetical order)

fanquake commented on doc/benchmarking.md:
````diff
@@ -19,19 +21,52 @@ After compiling bitcoin-core, the benchmarks can be run with:
 ---------------------
 
-    src/bench/bench_bitcoin --help
+```
````
I think having just `src/bench/bench_bitcoin -?` here is fine. Anyone wanting to see the options can run that, and what they see will be correct. Rather than copy-pasting blocks of output here, that will inevitably end up outdated.

jonatack replied:
Done.

fanquake commented on src/bench/bench.h:
```diff
@@ -18,16 +18,37 @@
+static void NameOfBenchmarkFunction(benchmark::Bench& bench)
 {
-    ... do any setup needed...
-    nanobench::Config().run([&] {
```
I think adding a note to look at nanobench.h for options is fine, but adding an arbitrary list of function signatures, most of which we don't use, seems like overkill (may be unlikely in the case of nanobench, but it's also just more opportunity for docs to diverge from the source). The user will likely have to look at the header anyways, to know what the options do.

jarolrod replied:
> it's also just more opportunity for docs to diverge from the source

that's a good point, lower cost of maintenance

jonatack replied:
Done.

jonatack commented:
Thanks @fanquake, @jarolrod and @kiminuo for the reviews. Updated per feedback and fixed the last commit message (bencharking -> benchmarking).

theStack commented:
Concept ACK, nice benchmark doc improvements!
Planning to verify the changes more in detail within the next days.

theStack reviewed (approved):
ACK d8513fe41102dcbfc05235f3b95e33eb1878f880 🚤 

Reviewed the documentation changes, ran `./src/bench/bench_bitcoin -filter="AddrManAdd|AddrManGetAddr|Base58CheckEncode|Base58Decode"` to verify that the example output format in the last commit matches.

za-kk commented:
ACK d8513fe41102dcbfc05235f3b95e33eb1878f880

Reviewed documentation and it's definitely an improvement to what was already there, also ran `./src/bench/bench_bitcoin -filter="AddrManAdd|AddrManGetAddr|Base58CheckEncode|Base58Decode"` and verified format matches with documentation 

Note - running `./src/bench/bench_bitcoin -?` causes an error in zsh as `?` is considered a wildcard, instead `./src/bench/bench_bitcoin "-?"` must be run. Now that we are specifying `-?` in the docs instead of `--help` could cause confusion when using zsh (which is default on macos). But I'm not sure whether the intricacies of different shells should impact the documentation 🤷
